### ISSUE TITLE
Position-dependent Bandpasses

### ISSUE THREAD

**Issue by kbarbary** [sncosmo/sncosmo#10]
There is currently no way to represent a filter whose transmission function depends on focal plane position, other than having many `Bandpass` objects to represent a single physical filter, e.g. `desg_0_0`, `desg_1_0`, etc.

The most natural solution is to implement a subclass of `Bandpass` that stores multiple transmission functions depending on position. We'll have to think about the details of the interface.

**Comment by dannygoldstein** [comment]
Volunteering to work on this.

**Comment by rbiswas4** [comment]
@dannygoldstein  Awesome! This would be very useful and get us up to speed on being able to analyze real data beyond average transmissions. 

Q1. Do you have a plan on @rcthomas design options:
> I think the design question here is whether to put a layer that generates Bandpass objects for any given x, y or make x, y an argument of a call to Bandpass

I tend to agree with making the bandpass layer approach (which also seems to be the @kbarbary's opinion. Has @kbarbary's opinion on this evolved? ), mostly because model methods only depend on the quantities like position through the bandpass. 

A possible case to keep in mind is to be able to assign transmission functions discreteley to ccds rather than a smooth interpolation in radius. 

Q2. Do you have a dataset in mind which you want to build as a test?

**Comment by kbarbary** [comment]
Still haven't thought of any designs that I *really* like, but this has got me thinking about the problem again.

I'd be interested to hear any ideas Danny has. It's a deep-reaching change since Bandpasses are used everywhere in sncosmo. There's the additional problem of trying to maintain backwards compatibility if possible.

**Comment by dannygoldstein** [comment]
The interface of `Bandpass` is set; rewriting it to implement `__call__` or `trans()` would require other core interfaces to be rewritten (via changes that percolate to `zpbandflux`, `bandflux`). This could affect backwards compatibility. 

If one does:

```python
>>> sncosmo.get_bandpass('megacam-i')
```

I think one should get back a subclass of `Bandpass` that looks something like:

```python
class DependentBandpass(Bandpass):

def at(self, x=None, y=None, ...):
      # compute wave, trans given x, y, etc. 
      return Bandpass(wave, trans)

@property
def trans(self):
    # return some average or characteristic transmission

@property
def wave(self):
    #return the wavelength that corresponds to trans
```

That way you can do something like:

```python
>>> sncosmo.get_bandpass('megacam-i').at(x=5., y=6.)
```

and get the bandpass you expect. But if you pass the `DependentBandpass` around you will still get the behavior you expect from methods like `bandflux` or `zpbandflux`, just with an "average" transmission. This is because `DependentBandpass` will implement the interface of `Bandpass`.

I was planning to do this for JLA. 

Danny

**Comment by dannygoldstein** [comment]
The parameters of the `at` method would vary from `DependentBandpass` to `DependentBandpass`, depending on the nature of each. You could use a metaclass to emit `DependentBandpass`es with different `at` methods as necessary.

Danny

**Comment by rcthomas** [comment]
Sounds like another vote for layer, though my opinion is that @kbarbary should rule on whether to implement as meta-class or just some other more boring factory method implementation.

Recently I've noticed in comments stuff about preserving backwards compatibility, little "regrets" @kbarbary has about the interface (capitalization and dashes in filter names for instance), and wanting to avoid large refactors in the code-base.  It's probably wise to just plan for doing some of those things for the next major (e.g. v2.0.0) release by deprecating and then break compatibility.  So for instance, you can go with the factory implementation here, but plan ahead to do the "deep-reaching" change that is probably less cumbersome fix baked into the interface.

**Comment by dannygoldstein** [comment]
Here is a candidate implementation.

```python
class DependentBandpass(Bandpass):
    """A bandpass in which the transmission depends on one or more
    parameters, such as focal plane position or chip ID."""

    def _coerce_bandpass(self, callable):
        """Coerce the result of callable (`wave`,`trans`) into a Bandpass
        object, with a new name based on the combination of self.name
        and the arguments passed to callable.

        """
        def _coerced(*args, **kwargs):
            wave, trans = callable(*args, **kwargs)
            name = self.name + '%s' % args
            return Bandpass(wave, trans, wave_unit=self.wave_unit,
                            trans_unit=self.trans_unit, name=name)
        return _coerced

    def __init__(self, callable, *characteristic_point,
                 wave_unit=u.AA, trans_unit=u.dimensionless_unscaled,
                 name=None):

        self.at = self._coerce_bandpass(callable)
        self.at.__doc__ = callable.__doc__
        self.characteristic_point = characteristic_point
        _cbp = self.at(*self.characteristic_point)
        wave = _cbp.wave
        trans = _cbp.trans
        super(DependentBandpass,
              self).__init__(wave, trans, wave_unit=self.wave_unit,
                             trans_unit=self.trans_unit, name=name)

    def __repr__(self):
        name = ''
        if self.name is not None:
            name = ' {0!r:s}'.format(self.name)
        return "<DependentBandpass{0:s} at 0x{1:x}>".format(name, id(self))
```

**Comment by kbarbary** [comment]
Semantically, I really like the idea of `sncosmo.get_bandpass('megacam-i').at(5., 6.)` returning a transmission function. It *sounds* right.

It's also nice that DependentBandpasses would still have `wave` and `trans` attributes, which would allow `model.bandflux('megacam-i', ...)` to work.

On the other hand one could argue that the user should have to explicitly say they want the transmission at the characteristic point, as otherwise it might not be clear that the bandpass is not uniform. That is, a `DependentBandpass` should only produce `Bandpass` objects, not act like one itself. Semantically this would make more sense, but it wouldn't have the previous nice attribute. Would have to have a separate category in the registry for these things.

How will `fit_lc`, `mcmc_lc` and `nest_lc` interact with these sorts of bandpasses? Presumably, the photometry data table will contain column(s) for position information, such as `'x'` and `'y'` or `'r'`. We may need some way to indicate which columns are position information, either in metadata or in a keyword argument. The `*_lc` functions would then have a block of code where, if there is position information, they'd do `band[i] = band[i].at(x[i], y[i])` for each observation to get normal `Bandpass` instances.

Regarding your implementation example: the actual implementation of the `callable` will be responsible for doing the actual interpolation work and will have to be a class itself in order to store data. Why not just have that class inherit from some parent class that implements the common stuff you have here? Then you can avoid the closure `_coerced` (unless you just really like closures?...). For example, say we're implementing a radially dependent bandpass:
```python
class NonUniformBandpass(Bandpass):
    # this is an abstract class.

class RadialBandpass(NonUniformBandpass):
    # implement `at` or some private method that `NonUniformBandpass.at` will call.
```

**Comment by dannygoldstein** [comment]
I like the idea of calling this class ` NonUniformBandpass` instead of a `DependentBandpass`. It's clearer. 

> On the other hand one could argue that the user should have to explicitly say they want the transmission at the characteristic point, as otherwise it might not be clear that the bandpass is not uniform. That is, a DependentBandpass should only produce Bandpass objects, not act like one itself. Semantically this would make more sense, but it wouldn't have the previous nice attribute. Would have to have a separate category in the registry for these things.

I agree with this, had the same thought. What do you think of the idea of the `wave` and `trans` attributes of a `NonUniformBandpass` being implemented as properties that raise exceptions when they are accessed? 

> Regarding your implementation example: the actual implementation of the callable will be responsible for doing the actual interpolation work and will have to be a class itself in order to store data. Why not just have that class inherit from some parent class that implements the common stuff you have here? Then you can avoid the closure _coerced (unless you just really like closures?...). For example, say we're implementing a radially dependent bandpass:

I don't think this is a good idea. What happens when you have a `RadialBandpass` that is also a `ChiptoChipBandpass`? You'll have to keep reimplementing the `at` function when you want to combine the `at` functions of two already implemented subclasses of `NonUniformBandpass`. You'll also create a huge class hierarchy. Not to mention, different nonuniform bandpasses will require different interpolation techniques depending on the peculiarities of the provided transmission data. I think it's better to take it on a case by case basis, include some good builtins, and leave the rest to the user by providing the interface to a `callable` that the user must implement to create a new `NonuniformBandpass`

**Comment by kbarbary** [comment]
> What do you think of the idea of the wave and trans attributes of a NonUniformBandpass being implemented as properties that raise exceptions when they are accessed?

That would be helpful for guiding users in the right direction, if we go this route. It's just a more helpful error message though, since you'd get an `AttributeError` exception anyway.

> I don't think this is a good idea. What happens when you have a RadialBandpass that is also a ChiptoChipBandpass? You'll have to keep reimplementing the at function when you want to combine the at functions of two already implemented subclasses of NonUniformBandpass.

Perhaps I'm missing your vision with the callables but it seems like you'll have to do this anyway if you use callables. The callables would have to be classes, and their `__call__` method would have to do basically the same thing as the `at` method, no? Would callables somehow allow you to trivially combine radial and chip-to-chip functionality?

**Comment by dannygoldstein** [comment]
> Perhaps I'm missing your vision with the callables but it seems like you'll have to do this anyway if you use callables. The callables would have to be classes, and their __call__ method would have to do basically the same thing as the at method, no? Would callables somehow allow you to trivially combine radial and chip-to-chip functionality?

Yes, you'll have to reimplement `at` for each `NonUniformBandpass` in both the callable scenario and the `RadialBandpass`/`ChiptoChipBandpass` (subclass) scenario that you propose. The main difference between the two is that in the subclass scenario, the same `at` is used to compute `Bandpass`es for all instances of a specific subclass (e.g. `RadialBandpass`), whereas in the callable scenario there is more flexibility to adapt `at` to peculiarities of the transmission data of a filter set that could be represented as a specific subclass. 

Take the SNLS / JLA (megacam) filter set as an example. In the following figure, light / dark corresponds to close / far away from the center of the focal plane.

![download-1](https://cloud.githubusercontent.com/assets/2769632/11102917/31337ace-8874-11e5-85f6-9eebb8b62082.png)

This filter set could be represented as a `RadialBandpass`. The transmission data are provided as a number of files giving the transmission as a function of wavelength at different radii on the focal plane. However, the wavelength sampling of the transmission function differs from file to file, making `sncosmo`'s standard 2D interpolation technique `RectBivariateSpline` unusable in this case. This creates some ambiguity wrt which interpolation technique to use to implement `at` for this dataset. One could try linear interpolation, but this causes a number of artifacts:

![download](https://cloud.githubusercontent.com/assets/2769632/11102899/0c2d985e-8874-11e5-899b-3aea2ac7999e.png)

A Gaussian Process produces better results (following is just z-band):

![download-2](https://cloud.githubusercontent.com/assets/2769632/11103164/a70a4d94-8875-11e5-9a31-b70352251d28.png)

but there is an nontrivial interpolation artifact at low and high wavelengths (following plot shows the residuals. The artifacts are also visible by eye in the above plot):

![download-3](https://cloud.githubusercontent.com/assets/2769632/11103234/0fbe1f5a-8876-11e5-81ed-f71a879707dd.png)

The downside of this approach is that storing the GP model may be costly. And it may take some time to evaluate. 

So basically how to implement `at` for this dataset is not totally clear, but it is clear that for optimal performance it will require a dataset-specific implementation. A different dataset that could be represented as a `RadialBandpass` may be better implemented as a `RectBivariateSpline` or via some other technique. The callable implementation of `NonUniformBandpass` allows for this intra-subclass variation. 

I think basically the user should just write a callable that takes the parameters that describe the nonuniformity of the transmission function (x, y, ccd, etc.) and return `wave` and `trans`, which can then be postprocessed into a real bandpass object with proper naming etc.

Danny

**Comment by kbarbary** [comment]
Sweet plots. Is that the new Viridis MPL colormap?

I see your point now. I agree that it is possible that every set of filters might need a different implementation, because we can't control how the data are sampled and handed to us. Perhaps you get my point already, but just to reiterate:

**The "callable" case:**

```python
# implement your callable:
class GPRadialBandpass(object):
    def __init__(...):
        ...
    def __call__(...):
        ...

# Use it:
mycallable = GPRadialBandpass(...)
band = NonUniformBandpass(mycallable)
```

**The inheritance case:**

```python
# implement you subclass
class GPRadialBandpass(NonUniformBandpass):
    def __init__(...):
        ...
    def _at(...):
        ...

# Use it:
band = GPRadialBandpass(...)
```

Practically speaking, the two options are very similar. `__call__` in the first option and `_at` in the second option could be identical. (In the second option, the superclass would implement `at` which would call `_at` and then create a `Bandpass` from the returned values). There are the same number of classes in each. Note that in the later case, the class "hierarchy" is still pretty flat - everything inherits from `NonUniformBandpass`; no reason to have deeper nested inheritance. I find the later easier to understand because it doesn't try to hide the fact that there's polymorphism going on by wrapping everything up in the same class. Tastes may vary though, I suppose.

**Comment by dannygoldstein** [comment]
No it's just a cubehelix palette. And I see the callable case more as:

```python
#implement your callable:
def _snls_bandfunc(x,y):
    # interpolate
    return wave, trans

# use it
band = NonUniformBandpass(_snls_bandfunc)
```

Maybe this is just a semantic difference.

**Comment by dannygoldstein** [comment]
And the above would be the kind of thing you would provide as a builtin.

**Comment by kbarbary** [comment]
But then the function `_snls_bandfunc()` would actually use some persistent data that would have to be instantiated outside the function as a global, right? You don't want to have to read in the bandpass definition from file(s) every time you call `_snls_bandfunc()`.

**Comment by dannygoldstein** [comment]
Yes, ok, so class SNLSBandFunc(object) then. 




Danny
